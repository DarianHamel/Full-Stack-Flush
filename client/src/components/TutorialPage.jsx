import React from "react";
import { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom"; // Import useNavigate
import axios from "axios";
import "../design/Tutorial.css";

const TutorialPage = () => {
  const [tutorials, setTutorials] = useState([]);
  const navigate = useNavigate(); // Initialize navigation
  const API_BASE_URL = "http://localhost:5050/api/tutorials"; 

  useEffect(() => {
    loadTutorials();
  }, []);

  const loadTutorials = async () => {
    try {
      const response = await axios.get(API_BASE_URL);
      setTutorials(response.data.tutorials);
    } catch (error) {
      console.error("Error fetching tutorials:", error);
    }
  };

  return (
    <div className="tutorials-container">
      <h1 className="tutorials-heading">Game Tutorials</h1>
      <div className="tutorials-list">
        {tutorials.map((tutorial) => (
          <Link key={tutorial._id} to={`/tutorial/${tutorial._id}`} className="tutorial-link">
            {tutorial.title}
          </Link>
        ))}
      </div>
    </div>
  );
};

export default TutorialPage;
