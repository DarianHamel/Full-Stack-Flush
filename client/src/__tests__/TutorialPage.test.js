import React from 'react';
import { render, screen, waitFor, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TutorialPage from '../components/TutorialPage';
import axios from 'axios';
import { MemoryRouter } from 'react-router-dom';


jest.mock('axios');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Link: ({ children, to, onClick }) => (
    <a 
      href={to} 
      onClick={(e) => {
        e.preventDefault(); 
        onClick?.(); 
      }}
    >
      {children}
    </a>
  ),
  useNavigate: () => jest.fn(), 
}));

describe('TutorialPage Component', () => {
  const mockUsername = 'testUser';
  const mockTutorials = [
    {
      _id: '1',
      title: 'Poker',
      content: 'Poker Basics',
      difficulty: 'Beginner',
      video_url: 'http://example.com/poker'
    },
    {
      _id: '2',
      title: 'Blackjack',
      content: 'Blackjack Basics',
      difficulty: 'Beginner',
      video_url: 'http://example.com/blackjack'
    }
  ];

  let user;

  beforeEach(() => {
    
    user = userEvent.setup();
    
    
    jest.clearAllMocks();
    
    
    axios.get.mockImplementation((url) => {
      if (url.includes('/api/tutorials')) {
        return Promise.resolve({ data: { tutorials: mockTutorials } });
      }
      if (url.includes('/api/user-tutorials')) {
        return Promise.resolve({ data: { tutorialsViewed: [] } });
      }
      return Promise.reject(new Error('Unknown URL'));
    });
    
    axios.post.mockResolvedValue({});
  });

  it('renders the component with heading and tutorial links', async () => {
    await act(async () => {
      render(
        <MemoryRouter>
          <TutorialPage username={mockUsername} />
        </MemoryRouter>
      );
    });

    
    expect(screen.getByText('Game Tutorials')).toBeInTheDocument();

    
    await waitFor(() => {
      expect(screen.getByText('Poker')).toBeInTheDocument();
      expect(screen.getByText('Blackjack')).toBeInTheDocument();
    });
  });

  it('loads and displays tutorials from the API', async () => {
    await act(async () => {
      render(
        <MemoryRouter>
          <TutorialPage username={mockUsername} />
        </MemoryRouter>
      );
    });

    
    expect(axios.get).toHaveBeenCalledWith('http://localhost:5050/api/tutorials');
    expect(axios.get).toHaveBeenCalledWith(`http://localhost:5050/api/user-tutorials/${mockUsername}`);

    
    await waitFor(() => {
      mockTutorials.forEach(tutorial => {
        expect(screen.getByText(tutorial.title)).toBeInTheDocument();
      });
    });
  });

  it('marks tutorial as viewed when clicked', async () => {
    await act(async () => {
      render(
        <MemoryRouter>
          <TutorialPage username={mockUsername} />
        </MemoryRouter>
      );
    });

    
    await waitFor(() => {
      expect(screen.getByText('Poker')).toBeInTheDocument();
    });

    
    await act(async () => {
      await user.click(screen.getByText('Poker'));
    });

    
    expect(axios.post).toHaveBeenCalledWith(
      `http://localhost:5050/api/user-tutorials/${mockUsername}`,
      { tutorialId: '1' }
    );
  });

  it('shows checkmark for viewed tutorials', async () => {
    
    axios.get.mockImplementation((url) => {
      if (url.includes('/api/tutorials')) {
        return Promise.resolve({ data: { tutorials: mockTutorials } });
      }
      if (url.includes('/api/user-tutorials')) {
        return Promise.resolve({ 
          data: { 
            tutorialsViewed: [{ _id: '1' }] 
          } 
        });
      }
      return Promise.reject(new Error('Unknown URL'));
    });

    await act(async () => {
      render(
        <MemoryRouter>
          <TutorialPage username={mockUsername} />
        </MemoryRouter>
      );
    });

    
    await waitFor(() => {
      expect(screen.getByText('Poker').closest('div')).toHaveTextContent('✔');
      expect(screen.getByText('Blackjack').closest('div')).not.toHaveTextContent('✔');
    });
  });

  it('handles API errors gracefully', async () => {
    
    axios.get.mockRejectedValue(new Error('API Error'));

    
    const consoleError = jest.spyOn(console, 'error').mockImplementation(() => {});

    await act(async () => {
      render(
        <MemoryRouter>
          <TutorialPage username={mockUsername} />
        </MemoryRouter>
      );
    });

    
    expect(consoleError).toHaveBeenCalledWith('Error fetching tutorials:', expect.any(Error));

    
    consoleError.mockRestore();
  });

  it('matches snapshot', async () => {
    let container;
    await act(async () => {
      container = render(
        <MemoryRouter>
          <TutorialPage username={mockUsername} />
        </MemoryRouter>
      ).container;
    });

    
    await waitFor(() => {
      expect(screen.getByText('Poker')).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });
});