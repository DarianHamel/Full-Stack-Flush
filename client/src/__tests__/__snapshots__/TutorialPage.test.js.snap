// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TutorialPage Component matches snapshot 1`] = `
<div>
  <div
    class="tutorials-container"
  >
    <h1
      class="tutorials-heading"
    >
      Game Tutorials
    </h1>
    <div
      class="tutorials-list"
    >
      <div
        class="tutorial-item"
      >
        <a
          href="/tutorial/1"
        >
          Poker
        </a>
      </div>
      <div
        class="tutorial-item"
      >
        <a
          href="/tutorial/2"
        >
          Blackjack
        </a>
      </div>
    </div>
  </div>
</div>
`;
